markAllBackingStoreVolatileFromTimer can stop the volatility timer too soon and not mark all surfaces as volatile.

mattwoodrow opened the pull request:
#### a6666b946bcc31afdea756373119ae86622de3c3
<pre>
markAllBackingStoreVolatileFromTimer can stop the volatility timer too soon and not mark all surfaces as volatile.
<a href="https://bugs.webkit.org/show_bug.cgi?id=264229">https://bugs.webkit.org/show_bug.cgi?id=264229</a>
&lt;<a href="https://rdar.apple.com/116711724">rdar://116711724</a>&gt;

Reviewed by Simon Fraser.

RemoteLayerWithRemoteRenderingBackingStoreCollection::markAllBackingStoreVolatileFromTimer stops the volatility timer if the operation is successful, since there should be no further backing stores to mark volatile.

This is an async operation though, and it&apos;s possible for markBackingStoreVolatileAfterReachabilityChange to have been called in the interim and there are now new backing stores that aren&apos;t yet volatile.

This change ensures that we reschedule the volatility timer if any markBackingStoreVolatileAfterReachabilityChange don&apos;t succesfully mark the relevant surfaces as volatile.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::markBackingStoreVolatileAfterReachabilityChange):

Canonical link: <a href="https://commits.webkit.org/270486@main">https://commits.webkit.org/270486@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ee92e021855f06aa5f106d59772854a134d23733

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24984 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26237 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27100 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22935 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/966 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23213 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25228 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2560 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21561 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27680 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2255 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28625 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22790 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22851 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26449 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2204 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/505 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3485 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6150 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2649 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2547 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
```diff
@@ -171,8 +171,10 @@
+    sendMarkBuffersVolatile(WTFMove(identifiers), [weakThis = WeakPtr { *this }](bool succeeded) {
         LOG_WITH_STREAM(RemoteLayerBuffers, stream << "RemoteLayerWithRemoteRenderingBackingStoreCollection::markBackingStoreVolatileAfterReachabilityChange - succeeded " << succeeded);
+        if (!succeeded && weakThis)
+            weakThis->scheduleVolatilityTimer();
```
Is there some risk that we never succeed and keep scheduling this timer forever?

mattwoodrow replied:
I think that risk is existing, the volatility timer is repeating, and it's only the successful marking of all surfaces that explicitly stops it.

mattwoodrow committed:
markAllBackingStoreVolatileFromTimer can stop the volatility timer too soon and not mark all surfaces as volatile.
https://bugs.webkit.org/show_bug.cgi?id=264229
<rdar://116711724>

Reviewed by Simon Fraser.

RemoteLayerWithRemoteRenderingBackingStoreCollection::markAllBackingStoreVolatileFromTimer stops the volatility timer if the operation is successful, since there should be no further backing stores to mark volatile.

This is an async operation though, and it's possible for markBackingStoreVolatileAfterReachabilityChange to have been called in the interim and there are now new backing stores that aren't yet volatile.

This change ensures that we reschedule the volatility timer if any markBackingStoreVolatileAfterReachabilityChange don't succesfully mark the relevant surfaces as volatile.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStoreCollection.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStoreCollection::markBackingStoreVolatileAfterReachabilityChange):

Canonical link: https://commits.webkit.org/270486@main
